Paper: PyTeCK: a Python-based automatic testing package for chemical kinetic models

kyleniemeyer opened the pull request:
Still (very much) finishing this paper, but wanted to get current progress submitted. Will have final draft committed tonight 👍

kyleniemeyer commented:
Alright, this is the actual complete draft... apologies for getting it in late.

kyleniemeyer commented:
Hmm, noticed that on the PDF site it seems like my paper isn't building... strange, it worked fine on my local machine

sbenthall commented:
Noted. @sarostru any idea what may be causing the build failure?

kyleniemeyer commented:
@bryanwweber suggested the build error might be due to the accent in "Zsély" in the LaTeX log file, although I don't know why that particular name (and not the others with accents) would cause a problem.

sarostru commented:
I don't know exactly, I'll be able to look into it tonight.  Sounds like maybe a missing font package or something.

katyhuff commented:
@kyleniemeyer Yeah... that's a bit weird. Your PR builds fine on my laptop. I'll review from that pdf. Maybe docutils needs to updated on the server?

kyleniemeyer commented:
Thanks! I've got some fixes I can push tomorrow, if you haven't started your review. 


> On Jun 24, 2016, at 1:07 AM, Katy Huff <notifications@github.com> wrote:
> 
> @kyleniemeyer Yeah... that's a bit weird. Your PR builds fine on my laptop. I'll review from that pdf. Maybe docutils needs to updated on the server?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

katyhuff commented:
Sounds good. I can certainly wait to read it on a plane that will take off in about 28 hours. So, I'll git pull before I take off and review anything that's in the repo at that point!

bryanwweber commented:
@katyhuff @sarostru I was unable to build it on my MacBook Pro with MacTeX 2016 and everything-updated conda Python. The error is here: https://github.com/scipy-conference/scipy_proceedings/blob/2016/publisher/build_paper.py#L146 when Python tries to decode the TeX log file, it fails, and if you look at the log file is has `\xe9` (or something like that) in a line describing an overfull hbox warning in the references, without the finishing byte. I'm not sure why TeX (I assume its TeX's fault since its in the log file) is not properly encoding the é that @kyleniemeyer mentioned.

kyleniemeyer commented:
OK, @katyhuff I'm more comfortable with the state of the paper now 😁.

Also def not sure why it wouldn't build, still no problems locally.

sarostru commented:
@bryanwweber thanks for investigating that was as far as I got as well, I wasn't able to resolve the problem beyond tracking it to a Tex problem.

At this point in the process we could replace the é with an e to get it building and then resolve the problem later.

kyleniemeyer commented:
OK, I just changed that to a regular "e" for now. Does that allow it to build?

sbenthall commented:
It appears to be building now. Thanks!

thewtex commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,935 @@
+
+Combustion simulations require chemical kinetic models to predict fuel
+oxidation, heat release, and pollutant emissions. These models are typically
+validated using qualitative—rather than quantitative comparisons—with limited
```
qualitative, rather than quantitative, comparisons

thewtex commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,935 @@
+In this paper, a data format is proposed for capturing results from
+experimental measurements of autoignition delay times. In addition,
+this paper describes the components of PyTeCK, a software package that
+quantities the performance of a chemical kinetic model in reproducing
```
quantities -> quantifies

thewtex commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,935 @@
+interpreting and converting between units, PyTables [Alted2002]_ for HDF5 file
+manipulation, Cantera [Goodwin2016]_ for chemical kinetics, and
+``pytest`` [Krekel2016]_ for unit testing. In addition, Travis-CI [Travis2016]_
+is used for continuous integration.
```
continuous integration testing.

thewtex commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,935 @@
+a volume-time history to capture nonideal pre- and post-ignition heat losses,
+as well as effects due to the compression stroke.
+This data can be provided with experimental ``datapoints`` in ChemKED as a list
+of lists, with the ``column`` index and units identifed:
```
identifed -> identified

thewtex commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,935 @@
+types, including laminar flames and flow reactors, building in visualization of
+results, and creating an open database of ChemKED files for experimental data.
+
+Acknowledgements
```
Acknowledgments

thewtex commented:
Very nice article!!

A few minor grammatical comments noted inline.

A few example time series plots would be interesting and some example output along with a reference for the code to produce the output would be nice.

kyleniemeyer commented:
Thanks @thewtex! I push commit 2b1d1ca that addressed all of your inline comments. I will also add some plots and example output, and a reference to a specific version of the code.

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+validated using qualitative, rather than quantitative, comparisons with limited
+sets of experimental data. Furthermore, while a plethora of published data exist
+for quantities of interest such as autoignition delay and laminar flame speed,
+most are not available in machine-readable formats. Such data is commonly
```
I found this confusing: CSV and to some extent Excel *are* machine-readable. Perhaps you mean something more like "standardized"?

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+ based on available experimental data.
+
+Fundamental combustion experiments typically study the behavior of fuels in
+idealized configuratinos at conditions relevant to applications in
```
``configuratinos`` -> ``configurations``

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+engines, and so ignition delay measurements provide useful validation measures
+for models aimed at capturing such phenomena.
+
+(Other combustion experimental measurements—such as extinction in perfectly
```
It seems odd to have a single sentence, single paragraph parenthetical here. I would suggest simply attaching it to the end of the preceding paragraph and leaving off the parenthesis.

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+
+Ignition delay times are typically measured with two categories of experiments:
+shock tubes and rapid compression machines.
+In shock tubes, discussed in detail in the literature [Chaos2010]_
```
these citations render a bit odd in-line.  Better to change to something like ``... in the literature, e.g. [Chaos2010]_ or [Hanson2014]_, a diaphragm ...``.  Or perhaps move it to the end of the sentence like `` ... In shock tubes, a diaphragm ... after a time delay.  These are discussed further in more detail in the literature, e.g. [Chaos2010]_ or [Hanson2014]_``

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+of a piston raises the temperature and pressure of a fuel/oxidizer
+mixture in a short period of time, after which ignition occurs.
+Shock tubes and rapid compression machines offer complementary approaches to
+measuring ignition delay times, as the former are typically limited to higher
```
not critical, but it might be nice to know what "high" and "low-to-intermediate" means in this context.  In my field "high temperature" means >~ 1 million K, but I'm guessing that's not true here ;)

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+Implementation of PyTeCK
+========================
+
+PyTeCK consists of three primary modules:
```
It might be worth a quick sentence or two pointing out that PyTeCK is packaged as a standard python package using setuptools.  That would make it clearer that you're using the word "modules" here to mean python modules, rather than the more general sense of "pieces of code".

kyleniemeyer replied:
I added this, as well as some installation instructions, via 206a337e025a098ab7dbd9fb720fd457740b3aa1

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+---------------------
+
+The PyTeCK module ``parse_files`` parses experimental data encoded in the
+newly proposed ChemKED (**chem**\ ical **k**\ inetics **e**\ xperimental
```
``newly proposed ... format ...`` -> ``format proposed by this paper`` or similar - i.e. get across the point that it's newly-proposed *here*

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+[Frenklach2007]_ [You2012]_ [PrIMe2016]_—but is written in YAML
+instead of XML. While XML is a powerful markup language, YAML offers a number
+of advantages: parsers and libraries exist for most programming langauges,
+it supports multiple data types and arrays, and YAML files are human-readable.
```
Bravo on this!  Very clear explanation on the value of YAML formats over XML.  (No change needed, just feedback)

kyleniemeyer replied:
I tried to emphasize this a bit more in db80022

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+``common-properties`` element is used for convenience.
+
+Modeling ignition in shock tubes or RCMs may require additional elements to
+capture nonidealities. Under certain conditions
```
Maybe ``nonidealities`` -> ``effects not accounted for by the simplest models``?   I suspect from context that "nonideality" means something specific in your field, but it's not clear to me what that means (although of course you're explaining it below more).

kyleniemeyer replied:
Good suggestion, thanks!

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+The PyTeCK ``tests`` directory [Niemeyer2016b]_ contains more examples of
+ChemKED files for shock tube and RCM experiments.
+
+The function ``parse_files.read_experiment`` takes a ChemKED-format file
```
You might want to show this as `` ` `parse_files.read_experiment()` `  `` (i.e. add the parenthesis) as per documentation conventions to make it more visibly apparent that this is a function.

eteq replied:
and same with *all* functions you explicitly mention later on by name

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+Pint [Grecco2016]_ interprets.
+Since many experimental datasets hold certain properties constant (e.g.,
+composition, pressure) while varying a single quantity (e.g., temperature),
+a ``common-properties`` element can desribe properties common to all
```
``desribe`` -> ``describe``

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+steps in parallel (if desired), as described in the next section. When running
+a simulation, PyTeCK creates an HDF5 file and opens it as a
+PyTables [Alted2002]_ table, then performs integration steps until it
+reaches desired end time (set as 100 times the experimental ignition delay):
```
reaches *the* desired end time (missing word)

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+PyTables [Alted2002]_ table, then performs integration steps until it
+reaches desired end time (set as 100 times the experimental ignition delay):
+
+.. code-block:: python
```
It's not clear what the purpose of this code block is.  It seems to simply demonstrate the internal workings of a simulation.  What's the purpose of showing this?  That needs to be spelled out more clearly, or the example removed.

kyleniemeyer replied:
I agree that there isn't much contribution to the paper to including this code block; the text already describes everything going ok. I just removed it (and added a ref to the integrator used internally by Cantera) via 15ca66c

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+internally ``nC7H16`` represents the species *n*\ -heptane, while other
+models may use ``C7H16``, ``C7H16-1``, or ``NXC7H16``, for example.
+PyTeCK has a preferred naming convention for some species (found in the
+``spec_key`` and ``spec_key_rev`` dictionaries in the ``utils.py`` module), but
```
This seems like insufficient information for figuring out what the preferred conventions are - it would be hard to understand what these conventions are without deep digging in the source code or similar.  Perhaps they should be listed explicitly here? (or better yet, in the docs as I suggest in the general comments)

kyleniemeyer replied:
Yeah, it wasn't particularly easy to get this information before. I can add the current species naming here if needed, but I also just added a new function `utils.print_species_names()` which prints the list of internal species identifiers (with common long names). This will also go into the PyTeCK docs, when written... 😄 

I added some text mentioning this function in 206a337e025a098ab7dbd9fb720fd457740b3aa1

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+need to match names of species in the ChemKED files.
+
+The ``model_variant_file`` YAML file is only needed in certain (uncommon) cases
+where the chemical kinetic model needs internal, manual changes for different
```
I found this discussion a bit confusing.  What does "manual changes" mean?  And is ``.cti`` a special format that provides this information?  A bit more background is needed on what's going on here (or maybe just a citation to something that explains more).

kyleniemeyer replied:
I've added some more detail on the CTI files, and tried to explain this (uncommon) model variant situation via a499821

eteq commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+After calculating the error associated with a dataset using Equation
+(:ref:`errorfunc`) and the overall error metric for a model using Equation
+(:ref:`overallerror`), ``evaluate_model`` prints the performance results
+to screen and saves them in a YAML file.
```
This seems a bit anti-climactic.  Wouldn't it make sense to return something that the user could potentially use in a programmatic manner rather than saving to disk?  It also seems odd to *require* printing to screen, as that's then difficult to combine in any sort of automated workflow (maybe a boolean switch?).

(for the paper, this is fine - it's an accurate description of the current situation.  Take this as a friendly suggestion for something you might change in the code)

kyleniemeyer replied:
Thanks for the suggestion—I went ahead and added to option for printing to screen as a command line option (`--print`), or option when calling `evaluate_model`. The function also returns the output dict (that is saved as a YAML file).

I updated the description of this behavior in 7e02405.

eteq commented:
@kyleniemeyer - I've reviewed this paper.  Note that this is quite far from my scientific field so I'm not terribly well-positioned to review the details of the model, but I tried to review from a more generic "scientific coding" perspective.  I've left a number of inline comments above, and here are a few more general comments:

## Related to this paper
* I suggest you spell out what PyTeCK actually means (i,e, "Python-based Test of Chemical Kinetic models") just once at the beginning to make it clear.
* I note that the actual code itself is only cited once via the ``Niemeyer2016b`` citation, and even then it references "PyTeCK version 0.1.0" which does not seem to actually exist.  When citing software (even/especially if it's what the whole paper is about) it's critical to point to a specific version so that the reader can tell what the paper refers to.  My suggestion for addressing this is to use a service like zenodo that will provide a DOI to cite a specific version of the code.  Even if this is a pre-release, it's much better citation practice.  Check out https://guides.github.com/activities/citable-code/ for instructions on how to do that.  It's also fine to have a direct link inline that points to the github repository in general (i.e., https://github.com/kyleniemeyer/PyTeCK) so that the reader has quicker access to the latest updates.

## Code-related

These are not strictly about the paper, but rather suggestions you might want to consider about the code itself, which came to me in the process of reviewing the paper.  Obviously better to prioritize the above comments for the paper itself above these, but if you like them you might also add them to the "future plans".

* Do you plan to publish PyTeCK on [PyPI](https://pypi.python.org/pypi)? This is valuable because it lets users ``pip install pyteck`` easily, among other things.  I see there are as yet no formal releases, so maybe you're just waiting for that.
* There are no install instructions.  Particularly given the lack of a PyPI entry, this might be useful.  I imagine it's as simple as ``python setup.py install`` from a quick glance, but still worth mentioning for those in your field who might not be particularly Python-savvy.
* It looks like you're planning this already, based on the "future plans" section, but I strongly recommend writing Sphinx docs for the code.  You've done a lot of work here and the paper explains everything very well, but it would be quite hard to actually use the code without some docs explaining the API in more detail and having some working examples.  Perhaps you can even adapt chunks of this paper into the documentation?  

That's all - with those suggestions in mind I can definitely recommend this for inclusion in the proceedings.  Nice work!

katyhuff commented:
Hi @kyleniemeyer -- you've gotten some great comments from other reviewers already. I agree that this paper will be a great addition to the proceedings and is in very good shape. Based on the review criteria, my review follows. Some minor comments will be added in-line. 

## Written Quality

The prose was written in excellent professional english and clearly communicates the narrative at a technical level appropriate for a general scipy audience. Highly domain-specific jargon was appropriately avoided or explained. 

## Technical Content

The technical content, describing a chemical kinetic model implementation for combustion, is sound and compelling. While an example demonstrating successful results generated by this software would improve this paper, the paper already succeeds at describing the model and implementation. 

## Novelty and Impact

The work seems to bring a more quantitative approach to model validations of this type and encourages more human and machine readable data formats.

## Verifiability

Software descriptions are accompanied by references to or examples of representative source code, the code and tests are available online, and data as well as software dependencies are appropriately cited.

## Other

All other miscellaneous requirements are satisfied. 

# Comments

Comments are provided inline.

katyhuff commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+   open-source Python-based package for automatic testing of chemical kinetic
+   models. Given a model of interest, PyTeCK automatically parses experimental
+   datasets encoded in a YAML format, validates the self-consistency of each
+   dataset, and performs simulations for each experimental datapoint. It then
```
No big deal, but ... is datapoint really one word?

kyleniemeyer replied:
Good question; seems like "data point" may be more standard. Updated in 9dba264

katyhuff commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+                           )
+
+where the ``validation.validate_gt`` function—borrowed heavily from Huff and
+Wang's PyRK [Huff2015]_ [Huff2015b]_—is
```
Glad this was useful! :)

katyhuff commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+mixture, :math:`e_i` is the internal energy of the *j*\ th species in mass
+units, :math:`v` is the specific volume of the mixture,
+and :math:`\dot{\omega}_i` is the overall molar production rate of the
+*i*\ th species. PyTeCK solves the system given by Equation (:ref:`systemodes`)
```
The data for this molar production rate... where does it come from? Is it input by the user or available in PyTECK as a library? Maybe I missed this explanation somewhere.

kyleniemeyer replied:
PyTeCK uses Cantera, which handles all of the actual chemical kinetics work. Basically, it creates a `cantera.IdealGasReactor` object for each simulation, which holds the thermochemical state information and has member functions for things like reaction rates, species molar production rates, etc. 

Placing the `IdealGasReactor` in a `cantera.ReactorNet` object then allows you to time-evolve the state based on the governing ODE system, either under constant volume conditions (the default) or with a predefined varying volume (handled by `VolumeProfile` or `PressureRiseProfile` objects). The `ReactorNet` uses implicit integration algorithms from SUNDIALS (CVODE, to be precise) to do that integration.

OK, so now to just incorporate that explanation in the text somehow... 😄

kyleniemeyer replied:
OK, I tried to add more explanation of how PyTeCK relies on Cantera earlier on via f7de2a2, and then the use of CVODE for integrating via 15ca66c.

Please let me know if that clarifies things!

kyleniemeyer commented:
Thanks @eteq and @katyhuff (and again to @thewtex)! I'll be responding to line comments that don't require any discussion with 🎉 to indicate that I've addressed them.

kyleniemeyer commented on papers/kyle_niemeyer/kyle_niemeyer.rst:
```diff
@@ -0,0 +1,936 @@
+information. Mandatory datapoint elements include the initial ``temperature``,
+``pressure``, and mixture ``composition``, as well as the experimental
+``ignition-delay`` and ``ignition-type`` (means by which PyTeCk detects
+ignition). All quantities provided include a magnitude and units, which
```
This is discussed in a bit more detail on page 5; is that explanation more clear? I have added "discussed in more detail later" here.

eteq commented:
Looking good, @kyleniemeyer . Did you give some thought to using Zenodo or otherwise pointing to a specific (reproducible by the reader) version of the code?

kyleniemeyer commented:
Thanks @eteq! At this point I believe I've responded to all of your line comments. As for the things you mentioned in your longer review:
 * I added the spelling out of PyTeCK with the first mention of the name in text via cca6c9f
 * I do plan to submit to Zenodo for a DOI—I've mainly been waiting to get to the point of an official release, that will correspond to the final version of this paper. (I'm also one of the people leading efforts to standardize [software citation practices](https://doi.org/10.7287/peerj.preprints.2169v2) so I'm fairly aware of the importance of citing a specific version 😛)
 * I also plan to submit to PyPI once the pre-release version is DOI'ed. For now, I've added some install instructions to the PyTeCK README, for using either `python setup.py install` or `pip install .`
 * I absolutely plan to write some Sphinx-based docs, and host them on Read the Docs. I hope to get this completed soon after this paper is finished.

I'm pretty sure I'm happy with the software for a pre-release version, so I should get a DOI and submitted to PyPI (with associated paper updates) shortly.

kyleniemeyer commented:
Hi @katyhuff—I believe I responded to your line comments, most of which weren't too tricky to resolve (I think). I do still plan on adding a short example at the end.

Did my changes (or responses) clarify the confusion about the ignition type and where molar production rates come from?

eteq commented:
Awesome, thanks @kyleniemeyer (both for those things and the software citation article ;)  I'm fully satisfied by this, so this looks good to go to me!

katyhuff commented:
Thanks! Looks great to me as well.

kyleniemeyer commented:
Thank you! The only things left before I consider it "done" are to add the appropriate citation with DOI for v0.1 of PyTeCK, and I also want to add a nice usage example 😄.

sarostru commented:
Maybe I don't need to post here, but for consistency :)

Due to the complication of the proceedings deadline lining up with everyone's travel schedule to attend the conference we've decided to allow authors to spend an additional 2 days, that is until July 12th to complete any major changes to the submission.  This may impact your enjoyment of the conference, so we hope not too many of you need to take the time.  For reviewers, we ask that you read over this final document before the conference is over and post your decision here in the comments on or before July 17th.

Thanks and have a great Scipy!
Your proceedings co-chairs Scott Rostrup (@sarostru) and Sebastian Benthall (@sbenthall)

kyleniemeyer commented:
Still working on the usage example—I'm trying to include something real rather than a toy problem. Should be ready tomorrow or Monday!

kyleniemeyer commented:
OK, I've pushed what I believe is the final draft of the paper. I've added a new usage example section (with figure!), as well as a DOI for version 0.1.0 of the software.

katyhuff commented:
:tada:

thewtex commented:
:+1: Well done!
